### DISCUSSION TITLE
Old beatmaps have different MD5 hash from online database

### DISCUSSION THREAD

**Discussion by Loreos7** [ppy/osu#19551] [Q&A] [answered]
https://user-images.githubusercontent.com/86934170/182387273-f4a9a3cf-d1aa-4ae4-8649-1017190a81c5.mp4

https://user-images.githubusercontent.com/86934170/182388323-a3413ab9-19af-4a3f-a330-945359c213a4.mp4

Trying to update:

https://user-images.githubusercontent.com/86934170/182388361-5539aa1d-1c2b-4c39-a335-66dab3487355.mp4

[runtime.log](https://github.com/ppy/osu/files/9242688/runtime.log)
[network.log](https://github.com/ppy/osu/files/9242689/network.log)
[database.log](https://github.com/ppy/osu/files/9242690/database.log)
[performance.log](https://github.com/ppy/osu/files/9242691/performance.log)
[updater.log](https://github.com/ppy/osu/files/9242692/updater.log)

https://osu.ppy.sh/beatmapsets/215#osu/712
https://osu.ppy.sh/beatmapsets/212#osu/2029
https://osu.ppy.sh/beatmapsets/304#osu/1304
https://osu.ppy.sh/beatmapsets/167#osu/398
https://osu.ppy.sh/beatmapsets/157#osu/352

**Comment by frenzibyte** [comment]
All attached beatmaps have an online hash that's different from the file's actual hash:
```csharp
  res.OnlineID
  712

  beatmapInfo.MD5Hash
  "d78a9f8b4e1eee0cfc57a97a866d289c"

  beatmapInfo.OnlineMD5Hash
  "cda78189103f008cf54b76346857240c"

  res.OnlineID
  2029

  beatmapInfo.MD5Hash
  "4f1d91c34f71a3de7f64fb3c06ec9a1f"

  beatmapInfo.OnlineMD5Hash
  "f3cdc44511df5dce54eff8ded4035bfc"

  res.OnlineID
  1304

  beatmapInfo.MD5Hash
  "934923e4218f4cf7113dc799ca5dea4d"

  beatmapInfo.OnlineMD5Hash
  "29dead15c6c36c61ba8a203e044e70d5"

  res.OnlineID
  398

  beatmapInfo.MD5Hash
  "f439f303808e1b07920037352184face"

  beatmapInfo.OnlineMD5Hash
  "983e34050ceec70385e32d87d707e101"

  res.OnlineID
  352

  beatmapInfo.MD5Hash
  "a8860ec5a4f61f667c054c948dab6741"

  beatmapInfo.OnlineMD5Hash
  "5a088c9c0a3ce1e98ccd27fd3f4fb047"

  res.OnlineID
  22333

  beatmapInfo.MD5Hash
  "8614b32eccd803fe0b5c676995e1e93e"

  beatmapInfo.OnlineMD5Hash
  "21b72797553a346cd96dbcd2e3895438"
```

cc/ @peppy @Ephemeralis

**Reply by Ephemeralis** [reply]
Triple digit beatmapsets. Holy shit. Wonder if this is related to a similar (but less severe) issue in https://github.com/ppy/osu/discussions/19495?

**Reply by peppy** [reply]
@Ephemeralis would be appreciated if you can check. The IDs are above (`/b/<OnlineID>` should work)

**Reply by frenzibyte** [reply]
I've updated the list to include one more beatmap coming from https://github.com/ppy/osu/discussions/19306#discussioncomment-3405933.

**Reply by peppy** [reply]
cc @Ephemeralis

**Reply by Ephemeralis** [reply]
I've reuploaded the updated versions of these beatmaps, save for ``167`` which currently fails BSS filesize bootstrapping due to a video file that is too large.

**Comment by frenzibyte** [comment]
Another one from https://github.com/ppy/osu/discussions/19495#discussioncomment-3496181:
```csharp
  res.OnlineID
  68597

  beatmapInfo.MD5Hash
  "e589fd1135a2edc85d984e4fd14c1fe2"

  beatmapInfo.OnlineMD5Hash
  "c0a6863a5fdd310ce05866466602b177"
```

**Comment by Loreos7** [comment]
Bunch of new ones I found @Ephemeralis 
https://osu.ppy.sh/beatmapsets/3#osu/91
https://osu.ppy.sh/beatmapsets/16#osu/119
https://osu.ppy.sh/beatmapsets/32#osu/102
https://osu.ppy.sh/beatmapsets/42#osu/122
https://osu.ppy.sh/beatmapsets/56#osu/164
https://osu.ppy.sh/beatmapsets/62#osu/200
https://osu.ppy.sh/beatmapsets/98#osu/218
https://osu.ppy.sh/beatmapsets/75#osu/275
https://osu.ppy.sh/beatmapsets/124#osu/277
https://osu.ppy.sh/beatmapsets/148#osu/333
https://osu.ppy.sh/beatmapsets/146#osu/331
https://osu.ppy.sh/beatmapsets/141#osu/313
https://osu.ppy.sh/beatmapsets/420#osu/2041

This one still goes with update button from the original post https://osu.ppy.sh/beatmapsets/167#osu/398
Other ones became available to set a record

**Reply by peppy** [reply]
I've already discussed automating checking these old beatmaps, so @Ephemeralis please hold off making any further manual changes. It's clear this issue isn't going to go away without applying fire.

**Reply by Ephemeralis** [reply]
Will be interesting to see what exactly is going on here if/when you figure it out.

**Comment by Kaerulans** [comment]
I've found some more maps that don't want to update, not sure if it's the same issue though since some of them are not as old:

https://osu.ppy.sh/beatmapsets/370819
https://osu.ppy.sh/beatmapsets/364282
https://osu.ppy.sh/beatmapsets/268377
https://osu.ppy.sh/beatmapsets/238692
https://osu.ppy.sh/beatmapsets/8992
https://osu.ppy.sh/beatmapsets/8976
https://osu.ppy.sh/beatmapsets/8898
https://osu.ppy.sh/beatmapsets/8651
https://osu.ppy.sh/beatmapsets/7151
https://osu.ppy.sh/beatmapsets/5770
https://osu.ppy.sh/beatmapsets/5170
https://osu.ppy.sh/beatmapsets/4581
https://osu.ppy.sh/beatmapsets/3862
https://osu.ppy.sh/beatmapsets/3

**Reply by smoogipoo** [reply]
Looking at the set https://osu.ppy.sh/beatmapsets/370819 in particular, the beatmap which doesn't match the online version is https://osu.ppy.sh/beatmapsets/370819#osu/820005.

| source | checksum |
| --- | --- |
| In-database | f3cfc011859c6f51845838b4d3ec3d68 |
| `osu.ppy.sh/osu/820005` | f3cfc011859c6f51845838b4d3ec3d68 |
| `online.db` | f3cfc011859c6f51845838b4d3ec3d68 |
| `/api/v2/beatmaps/lookup?id=820005` | f3cfc011859c6f51845838b4d3ec3d68 |
| Downloaded beatmap set | **af4a5fcdee1d876b35bc3edf7846ad85** |

The difference in these files:
![image](https://user-images.githubusercontent.com/1329837/196333517-c8d4c627-8560-41c9-b548-88c7af23cc4e.png)

**Reply by peppy** [reply]
Just to reiterate what I mentioned above in https://github.com/ppy/osu/discussions/19551#discussioncomment-3507108: this requires batch reprocessing and there's not much value checking them on a per-case basis. It is very clear there's a lot of outdated archives on S3.

**Reply by smoogipoo** [reply]
So it's the archives that are wrong, and not the individual `.osu` files? I ask because there are also checksums in the database that do not match _either_ the .osu or the archive, one example:
```
beatmap_id  db_hash                             file_hash
3805092     e3b4ecc0310207f3464f1adaed0794ea    ce44aaf3695473a6da25feb68f37e1aa
```
(57 such cases)

**Reply by peppy** [reply]
What does `db_hash` refer to here?

**Reply by smoogipoo** [reply]
`db_hash` is `osu_beatmaps.checksum`

Sorry, was wrong of me to put `online_hash` there - that one is the `.osu` file hash.

**Reply by peppy** [reply]
From where? Production?

**Reply by smoogipoo** [reply]
That's from the `data.ppy.sh` samples, yes.

**Reply by peppy** [reply]
There's only two possible values:

- downloading `.osu` from `https://old.ppy.sh/osu/3805092`
- downloading `.osz` from `/d/<set>`

Lazer uses the second, stable uses the first. The first is also matching to the database and therefore more correct. Some archives are not updated and need to be.

The case you mention, along with every other reported case, matches this pattern.

```
❯ md5 fromis_9\ -\ LOVE\ BOMB\ \(Cellina\)\ \[Easy\].osu
MD5 (fromis_9 - LOVE BOMB (Cellina) [Easy].osu) = ce44aaf3695473a6da25feb68f37e1aa
```

```csharp

mysql> select * from osu_beatmaps where beatmap_id = 3805092\G;
*************************** 1. row ***************************
      beatmap_id: 3805092
   beatmapset_id: 1849271
         user_id: 2490770
        filename: fromis_9 - LOVE BOMB (Cellina) [Easy].osu
        checksum: ce44aaf3695473a6da25feb68f37e1aa
         version: Easy
    total_length: 197
      hit_length: 176
      countTotal: 366
     countNormal: 88
     countSlider: 133
    countSpinner: 4
      diff_drain: 2
       diff_size: 2.6
    diff_overall: 2
   diff_approach: 3.5
        playmode: 0
        approved: 1
     last_update: 2022-10-01 13:02:37
difficultyrating: 1.75357
       playcount: 2440
       passcount: 1108
 youtube_preview: NULL
   score_version: 1
      deleted_at: NULL
             bpm: 154
1 row in set (0.00 sec)
```

All cases require the same fix, which is to reprocess all beatmaps and update `.osu` files in archives to match the raw ones.

I have no idea where you are getting `e3b4ecc0310207f3464f1adaed0794ea` from but it's definitely not on production. Maybe test with a beatmap that isn't actively being updated? It looks very recent.

**Reply by smoogipoo** [reply]
Right, I didn't notice the `last_update` column... Here's two more.

https://osu.ppy.sh/beatmapsets/18260#taiko/65313
```
database: 6b4551ebd46058dd31ff42565cf7d18c
osu: 7ce1ab636a0b118ee0c76cda4b2e1818
archive: 6b4551ebd46058dd31ff42565cf7d18c
```

https://osu.ppy.sh/beatmapsets/758986#fruits/1596618
```
database: 2c46f40a654e9741078219c0762d1cc7
osu: f101011c6e3cd26a987eb8309c992b09
archive: 2c46f40a654e9741078219c0762d1cc7
```

But I guess it follows the same pattern - so both the archive and database checksums are probably incorrect?

**Reply by peppy** [reply]
the `/osu/` file *has* to be the source of truth (ie. match the checksum) else stable will not allow score submission. looking at that second map, people are submitting scores so that is the case.

If `osu` in your last post mean the `.osu` file from the location I specified, then i tested the second and it matches the database / archive value you specify. Didn't check the first but probalyt the same.

**Reply by smoogipoo** [reply]
That's so weird, I guess my local files on that second map (https://osu.ppy.sh/beatmapsets/758986#fruits/1596618) were out of date. I can confirm `/osu/` matches, but I regularly refresh these files with the `data.ppy.sh` ones :/

However, in the case of https://osu.ppy.sh/beatmapsets/18260#taiko/65313, I do still get:
```
md5sum ~/Downloads/Masayoshi\ Minoshima\ feat.\ nomico\ -\ Bad\ Apple\!\!\ \(ouranhshc\)\ \[lepidon\!\ -\ Taiko\ Oni\].osu
7ce1ab636a0b118ee0c76cda4b2e1818  /home/smgi/Downloads/Masayoshi Minoshima feat. nomico - Bad Apple!! (ouranhshc) [lepidon! - Taiko Oni].osu
```
Can you double check that?

**Reply by peppy** [reply]
Yes, that one is broken completely. The fix is the same, which is to use the `.osu` from `/osu/` as the single source of truth. It will be applied in a single pass in the future.

**Comment by frenzibyte** [comment]
Another from https://github.com/ppy/osu/issues/21219:
- https://osu.ppy.sh/beatmapsets/940#osu/6903
- https://osu.ppy.sh/beatmapsets/1373#osu/15463

**Comment by chsungyesuzuki** [comment]
https://osu.ppy.sh/beatmapsets/3862 still has the problem
Also https://osu.ppy.sh/beatmapsets/7390

**Comment by Joehuu** [comment, marked as answer]
Fixed, see https://github.com/ppy/osu/issues/25078#issuecomment-1790107365.
